### PR TITLE
Add `pynvjitlink` as a dependency

### PR DESCRIPTION
**Pull request by brandon-b-miller**
This PR adds `pynvjitlink` as a hard dependency for cuDF. This should allow for MVC when launching numba kernels across minor versions of CUDA 12 up to the version of `nvjitlink` statically shipped with `pynvjitlink`. 

cc @bdice

### CODE REVIEW

**Comment by bdice** [comment]
@brandon-b-miller Here are some TODO items:

- [ ] Update cudf's conda recipe to depend on pynvjitlink (conda/recipes/cudf/meta.yaml)
- [ ] Update the error message if pynvjitlink isn't present?
- [ ] Add tests? Or are we covered already?

**Comment by brandon-b-miller** [comment]
> * [ ]  Add tests? Or are we covered already?

The _most_ thorough way of testing would be to add a CI job that has CUDA driver `12.X` and CTK `12.Y` where `Y>X`.  In this situation, _every_ numba kernel launched would require `pynvjitlink` to work, even outside of contexts where cuDF is involved at all. 

If we were to start building cuDF packages with CUDA `12.Y`, and _did not_ update to the `12.Y` driver in CI, then _certain_ cuDF tests would invoke the `pynvjitlink` codepath. If the driver, cuDF build version, and CTK all align in CI, the `pynvjitlink` codepath will not be tested.

**Comment by brandon-b-miller** [comment]
> Do we know what is needed for the devcontainer build?
> 
> Had looked at the log, but wasn't quite grasping what the error was

I think it's this:
```
  Could not solve for environment specs
  The following packages are incompatible
  ├─ cuda-version 12.0**  is requested and can be installed;
  └─ pynvjitlink is not installable because it requires
     └─ cuda-version >=12.2,<12.3 , which conflicts with any installable versions previously reported.
```

**Comment by jakirkham** [comment]
Thanks Brandon! 🙏

That's very helpful. Where did you find this in the logs?

Think this PR ( https://github.com/rapidsai/pynvjitlink/pull/45 ) should fix it

**Comment by brandon-b-miller** [comment]
> That's very helpful. Where did you find this in the logs?


In the [failing job](https://github.com/rapidsai/cudf/actions/runs/7546700596/job/20545020486?pr=14763), click the dropdown arrow next to the "Run build in devcontainer" job marked by the red x. From there, there's a smaller dropdown that can be clicked into through the small white triangle under the step marked "run command in container". This expands the log inside which I found the error.

**Comment by jakirkham** [comment]
Ah now I see. Thank you! 🙏

> ...the small white triangle under the step marked "run command in container"

This is what I was missing

**Comment by jakirkham** [comment]
The fix above is now in `pynvjitlink` version `0.1.10` with packages already published

Restarting the failed CI jobs. Let's see how things go

**Comment by jakirkham** [comment]
Looks like there was an issue with `pin_compatible`. Fixing as part of a `0.1.11` release of `pynvjitlink` in PR ( https://github.com/rapidsai/pynvjitlink/pull/47 ). Included results from `conda render` later in the PR to confirm we are getting the expected behavior

**Comment by jakirkham** [comment]
Ok with Bradley's help we no have `0.1.11`. Rerunning CI

**Comment by jakirkham** [comment]
It looks like `pynvjitlink` is now being installed 🎉

**Comment by jakirkham** [comment]
Ok looks like the wheel build has an issue

```python
ModuleNotFoundError: No module named 'scikit_build_core'
```

Note: Had to look at the raw logs (as the GHA GUI had some issue rendering)

**Comment by jakirkham** [comment]
Merging in latest from `branch-24.02` in case that helps

**Comment by jakirkham** [comment]
Failing due to timeout. Likely this issue ( https://github.com/conda/infrastructure/issues/869 )

Let's wait for that to clear up and retry the failing builds then

**Comment by brandon-b-miller** [comment]
Some failing testing jobs complaining that the module isn't installed. Hoping https://github.com/rapidsai/cudf/pull/14763/commits/d301cb98de7a2d3a1a363f04b2fb07c4c710530f sorts it out.

File: dependencies.yaml
Comment (bdice) [inline dependencies.yaml:622]: This isn't the right fix. This adds pynvjitlink as a testing dependency, rather than as a package dependency. We need this to appear in cudf's `pyproject.toml` for CUDA 12. Unfortunately, we don't support matrix entries for pyproject.toml yet (https://github.com/rapidsai/dependency-file-generator/issues/46). Therefore, the proper fix is to modify `ci/build_wheel.sh` and change one of the lines that deletes ptxcompiler to be a find-replace ptxcompiler with pynvjitlink. I will push a change for that.

Comment (brandon-b-miller) [reply dependencies.yaml:622]: I see, thanks for the context @bdice !

Comment (bdice) [reply dependencies.yaml:624]: Fixed in 8b8a12c2ba3b75008645090c3db856bb1b7518d9.

File: ci/build_wheel.sh
Comment (brandon-b-miller) [inline ci/build_wheel.sh:51]: I think this might be causing some problems with the pip installation:
```
ERROR: Could not find a version that satisfies the requirement pynvjitlink-cu12-cu12
```
should it be 
```suggestion
    sed -i "s/ptxcompiler/pynvjitlink/g" ${pyproject_file}
```
?

Comment (bdice) [reply ci/build_wheel.sh]: Yuck. There's a "ptxcompiler-cu12" being generated in the lines above, and then deleted. This is a hack but it'll be good enough, I guess. I adopted that fix and pushed.

**Comment by vyasr** [comment]
@brandon-b-miller @bdice I'm not sure how close this PR is to merging, but #14770 is needed in the interim. If you anticipate this merging today then we can probably wait. Otherwise let's merge the other PR and revert it as part of this changeset.

**Comment by brandon-b-miller** [comment]
To me the pieces seem to be in place for everything to work as expected fairly soon, although I have a murkier picture of some of the conda issues we've encountered on the pynvjitlink side.

**Comment by bdice** [comment]
@vyasr I'd like to finalize and merge this PR today. I will approve #14770 in the meantime but let's only merge if this PR is stalled.

**Comment by brandon-b-miller** [comment]
conda tests seem to be failing at collection time in CUDA 11.x with
```
E   ModuleNotFoundError: No module named 'pynvjitlink'
```
I think there's some extra logic needed here since `pynvjitlink` is a cuda 12.x dependency. I'll update.

**Comment by brandon-b-miller** [comment]
Hoping https://github.com/rapidsai/cudf/pull/14763/commits/92c6bb1c0f93f2328b26eba8d6126ed893d70f48 resolves the latest set of failures.

File: python/cudf/cudf/utils/_numba.py
Comment (bdice) [inline python/cudf/cudf/utils/_numba.py:132]: Can we update the comment above to mention pynvjitlink and the corresponding role of that package? This comment:
```
    # ptxcompiler is a requirement for cuda 11.x packages but not
    # cuda 12.x packages. However its version checking machinery
    # is still necessary. If a user happens to have ptxcompiler
    # in a cuda 12 environment, it's use for the purposes of
    # checking the driver and runtime versions is harmless
```

Comment (bdice) [reply python/cudf/cudf/utils/_numba.py:132]: @brandon-b-miller I would generally advocate reviewing this entire file and any other files that relate to ptxcompiler/pynvjitlink to make sure things are named sensibly, etc. in a way that will support both CUDA 11 and CUDA 12+. I want the code comments and docs to reflect the implemented design going forward.

Keep in mind that we don't want to name things "CUDA 12" in the code if we can avoid it if it is likely that later versions will act in the same way.

Comment (brandon-b-miller) [reply python/cudf/cudf/utils/_numba.py:134]: how about something like https://github.com/rapidsai/cudf/pull/14763/commits/7dbf9f22dbb63b6d33cfd9ad783dee60ec41555c ?

Comment (bdice) [reply python/cudf/cudf/utils/_numba.py:134]: > In a CUDA 12.x environment, ptxcompiler provides version checking, but not MVC directly

Is this true? We don't use ptxcompiler in CUDA 12 environments. No environment should have both ptxcompiler and pynvjitlink installed at the same time.

Comment (brandon-b-miller) [reply python/cudf/cudf/utils/_numba.py:134]: It's technically `_ptxcompiler.py` in this case - our slimmed down, vendored version of the few functions we need.

Comment (bdice) [reply python/cudf/cudf/utils/_numba.py:134]: Ooooo. But I don't know how to distinguish `ptxcompiler` the package (only used when on CUDA 11) from `_ptxcompiler.py` the internal helper file (always active) from the text of this comment. Documenting that kind of thing clearly is what I want to achieve before merging this.

Comment (brandon-b-miller) [reply python/cudf/cudf/utils/_numba.py:134]: some reworking in https://github.com/rapidsai/cudf/pull/14763/commits/e8a90b9ee26c8e365f5df8e705ed7e8c9f21681a

File: python/cudf/cudf/utils/_numba.py
Comment (bdice) [inline python/cudf/cudf/utils/_numba.py:108]:
```suggestion
    # Either ptxcompiler, or our vendored version (_ptxcompiler.py)
```

File: python/cudf/cudf/utils/_numba.py
Comment (bdice) [reply python/cudf/cudf/utils/_numba.py:134]: Much clearer! Thanks for iterating on this.

**Review by bdice — APPROVED** [review]
Approving with one typo fix that I will commit.

**Review by raydouglass — APPROVED** [review]

File: conda/recipes/cudf/meta.yaml
Comment (jakirkham) [inline conda/recipes/cudf/meta.yaml:101]: Once we have a clearer idea on intended compatibility ( https://github.com/rapidsai/pynvjitlink/issues/48 ), we may want to add some version constraints here

This could be done in a separate PR though

Comment (bdice) [reply conda/recipes/cudf/meta.yaml:101]: Yes, this is reasonable. John proposed `pynvjitlink >=0.1.11,<0.2.0a0` offline, which seems appropriate to me.

Comment (jakirkham) [reply conda/recipes/cudf/meta.yaml:101]: Yeah though let's discuss in the issue and we can do this as follow up (after this PR is merged)

**Comment by brandon-b-miller** [comment]
The merge is being blocked by what seems like unrelated issues building the libcudf docs
```
/__w/cudf/cudf/docs/cudf/source/libcudf_docs/api_docs/column_classes.rst:4: WARNING: Error when parsing function declaration.
If the function has no return type:
  Error in declarator or parameters-and-qualifiers
  Invalid C++ declaration: Expecting "(" in parameters-and-qualifiers. [error at 24]
    inline CUDF_HOST_DEVICE column_device_view slice (size_type offset, size_type size) const noexcept
    ------------------------^
```
This branch has the latest though, so it's possibly a problem on 24.02 - this ring any bells to anyone?

**Comment by vyasr** [comment]
Yeah this error probably comes from merging #13846 without it being fully up-to-date because some other PR merged bad docs changes. I'll take a look.

**Comment by vyasr** [comment]
Hoping that https://github.com/rapidsai/cudf/pull/14780 resolves this.

**Comment by jakirkham** [comment]
Thanks all! 🙏

**Pull request merged**
